Allocation size computation in calloc must not overflow

tautschnig opened the pull request:
calloc guarantees that the requested number of bytes is allocated, or
else NULL is returned. Since an overflowing multiplication would imply
that the requested size can never be allocated, NULL must be returned
(and we should not report an arithmetic overflow error).

Fixes: #4606

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Allocation size computation in calloc must not overflow

calloc guarantees that the requested number of bytes is allocated, or
else NULL is returned. Since an overflowing multiplication would imply
that the requested size can never be allocated, NULL must be returned
(and we should not report an arithmetic overflow error).

Fixes: #4606

tautschnig committed:
Support non-constant array sizes in zero initializer

We can use ARRAY_OF for those, just like we do for infinitely large
arrays.

Fixes: #4603

tautschnig committed:
Avoid out-of-memory conditions when zero-initializing objects

When code requests large arrays, we should not instantiate them, and
instead use ARRAY_OF.

Fixes: #4602

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e8d9053f26185d3853121e358b810f1aeec4dd03).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112196815
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 40b6fe7beef1d16fd8bf229c2d7a83813f444064).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112200459

kroening commented:
This has a sufficient but not necessary condition for overflow.
We do have an overflow predicate; expose that, and use that?

tautschnig committed:
Make overflow expressions available in C/C++ language front-ends

This enables testing for arithmetic overflow without generating an
arithmetic overflow.

tautschnig commented:
> We do have an overflow predicate; expose that, and use that?

Done!

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2811,6 +2811,46 @@ exprt c_typecheck_baset::do_special_functions(
+    identifier == CPROVER_PREFIX "overflow_shl" ||
+    identifier == CPROVER_PREFIX "overflow_unary_minus")
+  {
+    exprt overflow{identifier, typet{}, exprt::operandst{expr.arguments()}};
```
Why not pick some appropriate type? This seems likely to trip up `typecheck_expr_binary_arithmetic` eventually

tautschnig replied:
It's the job of the type checker to work out the type of expressions (based on the type of the operands) so I think this should be fine?

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d30df0eb11fe0ea96be40acb95bf21fa7004acc3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112285856
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
